lib/web_ui/lib/src/engine/text/canvas_paragraph.dart

```diff
@@ -205,6 +205,30 @@ class CanvasParagraph implements ui.Paragraph {
     return _layoutService.getPositionForOffset(offset);
   }
 
+  @override
+  ui.GlyphInfo? getClosestGlyphInfoForOffset(ui.Offset offset) => _layoutService.getClosestGlyphInfo(offset);
+
+  @override
+  ui.GlyphInfo? getGlyphInfoAt(int codeUnitOffset) {
+    final int? lineNumber = _findLine(codeUnitOffset, 0, numberOfLines);
+    if (lineNumber == null) {
+      return null;
+    }
+    final List<LayoutFragment> lineFragments = lines[lineNumber].fragments;
```

Comment:
I'm not sure if it's a good idea to calculate grapheme clusters only inside a single fragment. A grapheme cluster could start in one fragment, and end in another.

Here's an idea and let me know what you think:

1. Calculate grapheme clusters on the entire paragraph.
2. Find the grapheme cluster that contains `codeUnitOffset`.
3. Find the subset of fragments that intersect with that cluster (using `overlapsWith(...)`).
4. For each of these fragments, do `fragment.toTextBox(start: graphemeStart, end: graphemeEnd).toRect()`.
5. Calculate the union (?) of all the produced rects (using `Rect.expandToInclude(...)` maybe?).
6. For the direction, you can just use the direction from any of the intersecting fragments (I think all of them must have the same direction, because a grapheme cluster can't have multiple direction inside of it, right?)

Reply:
>  A grapheme cluster could start in one fragment, and end in another.

Ah interesting. When would this happen (so I can write a test for it)? When the codeunits that belong to the same grapheme are split into different TextSpans?

Reply:
> When the codeunits that belong to the same grapheme are split into different TextSpans?

Yes, that's the one possibility I could think of. Not sure if there are others.

Reply:
Alright I ended up doing grapheme breaks per line (all breaks within a grapheme cluster are prohibited so grapheme clusters won't be broken into different lines I think). PTAL.